lib/internal/worker/locks.js

```diff
@@ -0,0 +1,526 @@
+const {
+  SafeMap,
+  SymbolToStringTag,
+  Symbol,
+  BigUint64Array,
+  Int32Array,
+  Promise,
+  ObjectCreate,
+  PromiseResolve,
+  ObjectDefineProperties,
+  TypeError,
+} = primordials;
+
+const kName = Symbol('kName');
+const kMode = Symbol('kMode');
+
+// Class that contains all the state around active and pending locks.
+class LockInternals {
+  constructor() {
+    // We identify each thread by the pid + thread id combination.
+    this.clientId = `node-${process.pid}-${threadId}`;
+
+    // These maps store the resolve/reject functions for a) when the lock
+    // request is granted and b) when the final result of the operation is
+    // known.
+    this.availableMap = new SafeMap();
+    this.resolveMap = new SafeMap();
+    this.queueProcessingScheduled = false;
+
+    // Create an internal BroadcastChannel instance for communication with
+    // other threads.
+    this.channel = broadcastChannel('nodejs-internal:locking');
+    this.channel.on('message', (msg) => this.onmessage(msg));
+    // Post the message that will be sent when this thread closes down
+    // (regardless of how exactly that happens).
+    this.channel.postMessage({
+      op: 'terminated',
+      clientId: this.clientId,
+      threadId: threadId,
+      [onclose_message_symbol]: true
+    });
+    // By default, do not keep the event loop running just for locks.
+    this.channel.unref();
+
+    // Look for existing messages on the channel. There should be some, unless
+    // we are the first thread to set this up.
+    this.readSynchronousMessages();
+    if (!this.state) {
+      // This assertion should be valid because the first Worker instantiation
+      // on the main thread makes sure that locks will be initialized.
+      assert(isMainThread);
+      const sab = new SharedArrayBuffer(16);
+      this.state = {
+        // Use a shared global counter to get ids for the individual
+        // lock requests.
+        lockIdCounter: new BigUint64Array(sab, 0, 1),
+        // The mutex used to ensure that no two threads access the internal
+        // state at the same time. Contains -1 if unlocked, and the locker's
+        // threadId otherwise.
+        mutexOwner: new Int32Array(sab, 8, 1),
+        // Lists of the lock (request)s in their individual states.
+        held: [],
+        pending: [],
+        stolen: [],
+      };
+      this.unlockMutex();
+    }
+  }
+
+  readSynchronousMessages() {
+    let msg;
+    while (msg = receiveMessageOnPort(this.channel)) {
+      this.onmessage(msg.message);
+    }
+  }
+
+  publishState() {
+    this.channel.postMessage({
+      op: 'stateUpdate',
+      state: this.state,
+      [sticky_message_symbol]: true
+    });
+  }
+
+  lockMutex() {
+    // Lock the mutex and load all pending messages in order to update the
+    // internal state.
+    const mutex = this.state.mutexOwner;
+    let owner;
+    while ((owner = Atomics.compareExchange(mutex, 0, -1, threadId)) !== -1) {
```

Comment:
Ideally there would be less "magic numbers" and more constants here (they're codified in comments atm) but I'm not sure that'd actually make things clearer.

Reply:
I’ve replaced the `-1`s with `MUTEX_UNOWNED`, that should be a bit more expressive. Are there other constants that you’d like to see gone? :)